pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1775,31 +1866,56 @@ impl<'a> DatadirModification<'a> {
     /// All the modifications in this atomic update are stamped by the specified LSN.
     ///
     pub async fn commit(&mut self, ctx: &RequestContext) -> anyhow::Result<()> {
-        // Commit should never be called mid-wal-record
-        assert!(self.pending_zero_data_pages.is_empty());
-
         let mut writer = self.tline.writer().await;
 
         let pending_nblocks = self.pending_nblocks;
         self.pending_nblocks = 0;
 
         // Ordering: the items in this batch do not need to be in any global order, but values for
         // a particular Key must be in Lsn order relative to one another.  InMemoryLayer relies on
-        // this to do efficient updates to its index.
-        let mut write_batch = std::mem::take(&mut self.pending_data_pages);
+        // this to do efficient updates to its index.  See [`wal_decoder::serialized_batch`] for
+        // more details.
 
-        write_batch.extend(
-            self.pending_metadata_pages
+        let metadata_batch = {
+            let pending_meta = self
+                .pending_metadata_pages
                 .drain()
                 .flat_map(|(key, values)| {
                     values
                         .into_iter()
                         .map(move |(lsn, value_size, value)| (key, lsn, value_size, value))
-                }),
-        );
+                })
+                .collect::<Vec<_>>();
+
+            if pending_meta.is_empty() {
+                None
+            } else {
+                Some(SerializedValueBatch::from_values(pending_meta))
+            }
+        };
+
+        let data_batch = self.pending_data_batch.take();
+
+        let maybe_batch = match (data_batch, metadata_batch) {
+            (Some(mut data), Some(metadata)) => {
+                data.extend(metadata);
```

Comment:
This seems to be carry-over from the old code, and we have (debug) assertions for it, but why is it safe to order the metadata batch after the data batch? Will there never be conflicting keys in the two?

Reply:
> Will there never be conflicting keys in the two?

That's right. See `DatadirModification::is_data_key`.

"Data keys" are relational block keys and SLRU keys.
SLRU keys are not something that PG writes to directly, but we infer the updates for them. However, PG may
request to read both directly.

"Metadata keys" are pageserver specific keys. They are not in PG's vocabulary and are never requested.

> but why is it safe to order the metadata batch after the data batch?

The more generic answer is that `SerializedValueBatch` only ordering constraint is that metadata entries for any given key are ordered by LSN. Couple this with the fact that the WAL stream can't go backwards and that's why it's safe to extend here.

Reply:
Makes sense, thanks for the explanation!